### ISSUE TITLE
GraphQL client research

### ISSUE THREAD

**Issue by krzysztofwolski** [saleor/apps#230]
Known issues to be addressed with the current setup: 
- when using generated queries types have to be passed manually to the function. Ergonomics could be improved:
```
const saleorVersion = await client
  .query<SaleorVersionQuery>(SaleorVersion, {})
  .toPromise()
  .then((res) => {
    return res.data?.shop.version;
  });
```
- all the generated code is placed in a single file. It's size makes hard to follow the changes of operations/setup (had this issue during investigating webhooks fragments issue)

Initial proposal:
- use a new client-preset https://the-guild.dev/graphql/codegen/plugins/presets/preset-client
- drop URQL since we are not using any of its advanced features in favor of graphql-request (simpler to use and smaller client bundle size impact)
- add react-query for it's great hooks interface/cache/state management. We are already using this package with apps (because of tRPC). Having a unified remote data interface would be a great win for consistency
- investigate if splitting file is a viable option

Expected spike outcome:
- migrate one of the existing apps with client and server-side requests to the new setup
- any custom functions should be extracted to a shared package for faster adoption in monorepo
- check the impact on client bundle size

**Comment by lkostrowski** [comment]
Hey @mmiszy any thoughts on that?

**Comment by krzysztofwolski** [comment]
@mmiszy love to hear about any issues with URQL you had so far. Any thoughts about keeping queries inside separate `.gql` files vs `.ts[x]` near the rest of the code?

**Comment by lkostrowski** [comment]
My 3 cents:
1. I like to have gql colocated near its usage
2. Fragments I usually have in `graphql/fragements`, if they are shared across queries
3. I'd like to have generate, typed functions. Like we have `useSomeQuery`, id like to have something like this but without react

**Comment by mmiszy** [comment]
I don't recall any specific issues with GraphQL.

Does react-query also normalize the data from GraphQL? That's probably the biggest benefit of using urql or apollo.

I think we'd rather have all queries and mutations in `.graphql` files – better code completion and ability to run them from IDE.

**Comment by krzysztofwolski** [comment]
Most of our apps run queries on the backend side, so we did not have big benefits from denormalization. But that's a good point, before switching the library I'll check what can be improved in urql setup :+1:

**Comment by lkostrowski** [comment]
>I think we'd rather have all queries and mutations in .graphql files – better code completion and ability to run them from IDE.

Can't it be achieved in VSC? In Jetbrains IDE inline graphql is supported very well

**Comment by mmiszy** [comment]
Also, colocating near usage sounds great and all but we have to keep in mind that queries' and fragments' names have to be globally unique. It's easier to achieve this when all GQL files are in a single directory.
